src/Nethermind/Nethermind.Consensus.AuRa/Validators/ReportingContractBasedValidator.PersistentReports.cs

```diff
@@ -0,0 +1,181 @@
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU Lesser General Public License for more details.
+// 
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+// 
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Nethermind.Consensus.Transactions;
+using Nethermind.Core;
+using Nethermind.Dirichlet.Numerics;
+
+namespace Nethermind.Consensus.AuRa.Validators
+{
+    public partial class ReportingContractBasedValidator : ITxSource
+    {
+        /// <summary>
+        /// The maximum number of reports to keep queued
+        /// </summary>
+        internal const int MaxQueuedReports = 10;
+        
+        /// <summary>
+        /// The maximum number of malice reports to include when creating a new block.
+        /// </summary>
+        internal const int MaxReportsPerBlock = 10;
+        
+        /// <summary>
+        /// Don't re-send malice reports every block. Skip this many before retrying.
+        /// </summary>
+        private const int ReportsSkipBlocks = 1;
+
+        private readonly LinkedList<PersistentReport> _persistentReports;
+        private long _resentReportsInBlock = 0;
+        
+        public IEnumerable<Transaction> GetTransactions(BlockHeader parent, long gasLimit)
+        {
+            foreach (var transaction in _contractValidator.GetTransactions(parent, gasLimit))
+            {
+                if (gasLimit >= transaction.GasLimit)
+                {
+                    gasLimit -= transaction.GasLimit;
+                    yield return transaction;
+                }
+            }
+
+            if (_contractValidator.ForSealing && IsPosdao(parent.Number + 1))
+            {
+                FilterReports(parent);
+                foreach (var persistentReport in _persistentReports.Take(MaxReportsPerBlock))
+                {
+                    var transaction = ValidatorContract.ReportMalicious(persistentReport.ValidatorAddress, persistentReport.BlockNumber, persistentReport.Proof);
+                    if (transaction != null && gasLimit >= transaction.GasLimit)
+                    {
+                        gasLimit -= transaction.GasLimit;
+                        yield return transaction;
+                    }
+                }
+            }
+        }
+        
+        private void ResendPersistedReports(BlockHeader blockHeader)
+        {
+            var blockNumber = blockHeader.Number;
+            if (!IsPosdao(blockNumber))
+            {
+                if (_logger.IsTrace) _logger.Trace("Skipping resending of queued malicious behavior reports.");
+            }
+            else
+            {
+                FilterReports(blockHeader);
+                TruncateReports();
+
+                if (blockNumber > _resentReportsInBlock + ReportsSkipBlocks)
```

Comment:
are you capable of reporting a malicious signer in block 1? (edge case)

Reply:
Yes it can happen (like 2 blocks in same step from one validator)